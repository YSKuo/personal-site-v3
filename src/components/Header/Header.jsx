import React from "react";
import styled from "styled-components";
import MenuIcon from "@material-ui/icons/Menu";
import {
  AppBar,
  Button,
  Link,
  Slide,
  useScrollTrigger,
  IconButton,
  Hidden,
  List,
  ListItem,
  Drawer,
  Container,
  Toolbar,
} from "@material-ui/core";
import { MEDIA_QUERY_SM, MEDIA_QUERY_MD } from "../../constants/breakpoint";

const StyledAppBar = styled(AppBar)`
  border-bottom: 1px solid #dbdbdb;
  position: fixed;
  min-width: 100vw;
  top: 0;
  left: 0;
  right: 0;
  z-index: 1000;
  transition: 0.1s ease-in;
`;

const StyledToolbar = styled(Toolbar)`
  justify-content: space-between;
`;

const NavLink = styled(Button)`
  font-size: 1.2rem;

  & ~ & {
    margin-left: 1rem;
  }
`;

function HideOnScroll(props) {
  const { children, window } = props;
  const trigger = useScrollTrigger({ target: window ? window() : undefined });

  return <Slide in={!trigger}>{children}</Slide>;
}

function isTabActive(link) {
  return link.url === "/"
    ? location.pathname === link.url
    : location.pathname.includes(link.url);
}

function Header({ config, theme }) {
  const { siteTitle, tabLinks } = config;
  const [state, setState] = React.useState(false);

  const toggleDrawer = () => {
    setState(!state);
  };

  return (
    <>
      <HideOnScroll>
        <StyledAppBar color="primary">
          <Container>
            <StyledToolbar component="nav" disableGutters>
              <Link color="textPrimary" variant="h5" href="/" underline="none">
                {siteTitle}
              </Link>

              <Hidden smUp>
                <IconButton onClick={toggleDrawer}>
                  <MenuIcon />
                </IconButton>
                <Drawer anchor="left" open={state} onClose={toggleDrawer}>
                  {tabLinks &&
                    tabLinks.map((link) => (
                      <ListItem button key={link.title}>
                        <Button
                          color={isTabActive(link) ? "secondary" : "default"}
                          variant={isTabActive(link) ? "contained" : "default"}
                          href={link.url}
                        >
                          {link.title}
                        </Button>
                      </ListItem>
                    ))}
                </Drawer>
              </Hidden>

              <Hidden xsDown>
                <Toolbar disableGutters>
                  <List>
                    {tabLinks &&
                      tabLinks.map((link) => {
                        return (
                          <NavLink
                            color={isTabActive(link) ? "secondary" : "default"}
                            key={link.title}
                            href={link.url}
                          >
                            {link.title}
                          </NavLink>
                        );
                      })}
                  </List>
                </Toolbar>
              </Hidden>
            </StyledToolbar>
          </Container>
        </StyledAppBar>
      </HideOnScroll>
    </>
  );
}

export default Header;
