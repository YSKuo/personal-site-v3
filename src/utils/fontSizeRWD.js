import { fontSizes } from "../constants/font";
import { mediaQueryBreakpoint } from "../constants/breakpoint";

function fontSizeRWD(text, [xs, sm, md]) {
  return `${text} {
    ${mediaQueryBreakpoint("xs")} {
      font-size: ${fontSizes[xs]};
    }
    ${mediaQueryBreakpoint("sm")} {
      font-size: ${fontSizes[sm]};
    }
    ${mediaQueryBreakpoint("md")} {
      font-size: ${fontSizes[md]};
    }
  }`;
}

export default fontSizeRWD;
