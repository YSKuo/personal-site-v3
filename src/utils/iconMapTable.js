import React from "react";

import InstagramIcon from "@material-ui/icons/Instagram";
import LinkedInIcon from "@material-ui/icons/LinkedIn";
import GitHubIcon from "@material-ui/icons/GitHub";
import EmailIcon from "@material-ui/icons/Email";
import TwitterIcon from "@material-ui/icons/Twitter";

const iconMapTable = {
  Email: <EmailIcon />,
  Twitter: <TwitterIcon />,
  GitHub: <GitHubIcon />,
  Instagram: <InstagramIcon />,
  LinkedIn: <LinkedInIcon />,
};

export default iconMapTable;
