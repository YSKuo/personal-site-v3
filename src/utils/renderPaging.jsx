import React from "react";
import styled from "styled-components";
import { Grid, Button, IconButton, Hidden } from "@material-ui/core";
import ArrowBackIosIcon from "@material-ui/icons/ArrowBackIos";
import ArrowForwardIosIcon from "@material-ui/icons/ArrowForwardIos";
import MoreHorizIcon from "@material-ui/icons/MoreHoriz";

function renderPaging(pageName, pageContext, location) {
  const { currentPageNum, pageCount } = pageContext;
  const prevPage =
    currentPageNum - 1 === 1
      ? `/${pageName}`
      : `/${pageName}/${currentPageNum - 1}/`;
  const nextPage = `/${pageName}/${currentPageNum + 1}/`;
  const isFirstPage = currentPageNum === 1;
  const isLastPage = currentPageNum === pageCount;
  const currentPage = location.pathname.match(/[0-9]+/)
    ? Number(location.pathname.match(/[0-9]+/))
    : 1;

  return (
    <Grid container component="nav" alignItems="center">
      <PaginationPreviousButton item xs={6} sm={2}>
        {!isFirstPage && (
          <Button
            href={prevPage}
            color="primary"
            variant="outlined"
            startIcon={<ArrowBackIosIcon />}
          >
            Previous
          </Button>
        )}
      </PaginationPreviousButton>

      <Hidden xsDown>
        <Grid item sm={8}>
          <PaginationList>
            {[...Array(pageCount)].map((_val, index) => {
              const pageNum = index + 1;
              if (
                pageNum === 1 ||
                pageNum === pageCount ||
                pageNum === currentPage ||
                pageNum === currentPage + 1 ||
                pageNum === currentPage - 1
              ) {
                return (
                  <PaginationNumber key={`${pageName}-page-${pageNum}`}>
                    <PaginationLink
                      variant={pageNum === currentPage ? "contained" : "text"}
                      size={pageNum === currentPage ? "medium" : "small"}
                      color={pageNum === currentPage ? "secondary" : "default"}
                      href={
                        pageNum === 1
                          ? `/${pageName}`
                          : `/${pageName}/${pageNum}/`
                      }
                      disableElevation
                    >
                      {pageNum}
                    </PaginationLink>
                  </PaginationNumber>
                );
              }
              if (pageNum === currentPage + 2 || pageNum === currentPage - 2) {
                return (
                  <PaginationNumber key={`${pageName}-page-${pageNum}`}>
                    <IconButton title="ellipsis" disabled>
                      <MoreHorizIcon />
                    </IconButton>
                  </PaginationNumber>
                );
              }
            })}
          </PaginationList>
        </Grid>
      </Hidden>
      <PaginationNextButton item xs={6} sm={2}>
        {!isLastPage && (
          <Button
            href={nextPage}
            color="primary"
            variant="outlined"
            endIcon={<ArrowForwardIosIcon />}
          >
            Next
          </Button>
        )}
      </PaginationNextButton>
    </Grid>
  );
}

export default renderPaging;

const PaginationList = styled.ul`
  display: flex;
  text-align: center;
  justify-content: center;
  align-items: center;
  list-style: none;
`;
const PaginationNumber = styled.li``;
const PaginationLink = styled(Button)`
  height: 2rem;
  width: 2rem;
  border-radius: 50%;
  min-width: initial;
  padding: 0 0.5rem;
  margin: 0 1rem;
`;

const PaginationPreviousButton = styled(Grid)`
  text-align: start;
`;

const PaginationNextButton = styled(Grid)`
  text-align: end;
`;
