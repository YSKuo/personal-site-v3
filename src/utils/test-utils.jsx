import React from "react";
import { ThemeProvider, createGlobalStyle } from "styled-components";
import { createMuiTheme, MuiThemeProvider } from "@material-ui/core/styles";
import { render } from "@testing-library/react";
import { theme } from "../constants/theme";

const SizeWrapper = ({ children }) => {
  const initTheme = createMuiTheme({
    props: { MuiWithWidth: { initialWidth: "lg" } },
  });

  return (
    <MuiThemeProvider theme={initTheme}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </MuiThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: SizeWrapper, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render, SizeWrapper };
