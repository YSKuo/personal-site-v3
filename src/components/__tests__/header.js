import React from "react";
import userEvent from "@testing-library/user-event";

import Header from "../Header";
import config from "../../../data/SiteConfig";
import { render } from "../../utils/test-utils";

function renderHeader(props) {
  const utils = render(<Header config={config} />);

  const header = utils.getByRole("banner");
  return { ...utils, header };
}

test("Header snapshot at lg screen width", () => {
  const { header } = renderHeader();
  expect(header).toMatchSnapshot();
});

test("Header is revealed", () => {
  const { header } = renderHeader();
  expect(header).toBeDefined();
});
