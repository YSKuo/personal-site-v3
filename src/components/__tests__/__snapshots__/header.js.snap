// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header snapshot at lg screen width 1`] = `
<header
  class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary sc-gsTCUz lbsyTM mui-fixed MuiPaper-elevation4"
  style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <nav
      class="MuiToolbar-root MuiToolbar-regular sc-dlfnbm cZnpQH"
    >
      <a
        class="MuiTypography-root MuiLink-root MuiLink-underlineNone sc-bdfBwQ HrXuG MuiTypography-h5 MuiTypography-colorTextPrimary"
        href="/"
      >
        Arsene's Alibi
      </a>
      <div
        class="MuiToolbar-root MuiToolbar-regular"
      >
        <a
          aria-disabled="false"
          class="MuiButtonBase-root MuiButton-root MuiButton-text sc-hKgILt xTIhp"
          href="/"
          tabindex="0"
        >
          <span
            class="MuiButton-label"
          >
            Home
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
        <a
          aria-disabled="false"
          class="MuiButtonBase-root MuiButton-root MuiButton-text sc-hKgILt blgYtI"
          href="/about"
          tabindex="0"
        >
          <span
            class="MuiButton-label"
          >
            About
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
        <a
          aria-disabled="false"
          class="MuiButtonBase-root MuiButton-root MuiButton-text sc-hKgILt blgYtI"
          href="/blog"
          tabindex="0"
        >
          <span
            class="MuiButton-label"
          >
            Blog
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
      </div>
    </nav>
  </div>
</header>
`;
